src/datasources/accounts/counterfactual-safes/counterfactual-safes.datasource.spec.ts

```diff
@@ -0,0 +1,407 @@
+    await migrator.migrate();
+    mockConfigurationService.getOrThrow.mockImplementation((key) => {
+      if (key === 'expirationTimeInSeconds.default') return faker.number.int();
+    });
+
+    target = new CounterfactualSafesDatasource(
+      fakeCacheService,
+      sql,
+      mockLoggingService,
+      mockConfigurationService,
+    );
+  });
+
+  afterEach(async () => {
+    await sql`TRUNCATE TABLE accounts, account_data_settings, counterfactual_safes CASCADE`;
+    fakeCacheService.clear();
+    jest.clearAllMocks();
+  });
+
+  afterAll(async () => {
+    await testDbFactory.destroyTestDatabase(sql);
+  });
+
+  describe('createCounterfactualSafe', () => {
+    it('should create a Counterfactual Safe', async () => {
+      const address = getAddress(faker.finance.ethereumAddress());
+      const [account] = await sql<
+        Account[]
+      >`INSERT INTO accounts (address) VALUES (${address}) RETURNING *`;
+      const createCounterfactualSafeDto =
+        createCounterfactualSafeDtoBuilder().build();
+
+      const actual = await target.createCounterfactualSafe(
+        account,
+        createCounterfactualSafeDto,
+      );
+      expect(actual).toStrictEqual(
+        expect.objectContaining({
+          id: expect.any(Number),
+          chain_id: createCounterfactualSafeDto.chainId,
+          creator: account.address,
+          fallback_handler: createCounterfactualSafeDto.fallbackHandler,
+          owners: createCounterfactualSafeDto.owners,
+          predicted_address: createCounterfactualSafeDto.predictedAddress,
+          salt_nonce: createCounterfactualSafeDto.saltNonce,
+          singleton_address: createCounterfactualSafeDto.singletonAddress,
+          threshold: createCounterfactualSafeDto.threshold,
+          account_id: account.id,
+        }),
+      );
+    });
+
+    it('should delete the cache for the account Counterfactual Safes', async () => {
+      const address = getAddress(faker.finance.ethereumAddress());
+      const [account] = await sql<
+        Account[]
+      >`INSERT INTO accounts (address) VALUES (${address}) RETURNING *`;
+      await target.createCounterfactualSafe(
+        account,
+        createCounterfactualSafeDtoBuilder().build(),
+      );
+      await target.getCounterfactualSafesForAccount(account);
+
+      // check CF Safes for the address are in the cache
+      const cacheDir = new CacheDir(`counterfactual_safes_${address}`, '');
+      const cacheContent = await fakeCacheService.get(cacheDir);
+      expect(JSON.parse(cacheContent as string)).toHaveLength(1);
+
+      // the cache is cleared after creating a new CF Safe for the same account
+      await target.createCounterfactualSafe(
+        account,
+        createCounterfactualSafeDtoBuilder().build(),
+      );
+      const afterCreation = await fakeCacheService.get(cacheDir);
+      expect(afterCreation).toBeUndefined();
+    });
+
+    // TODO: move this check to the repository.
```

Comment:
I've also left this here for reference but this test should be moved to the Repository when it's implemented.